File: 15-Dynamic-Programming/08-Digit-DP/java/DigitDP.java
public class DigitDP { public static int findIntegers(int n){if(n==0)return 1;int[] fib=new int[32];fib[0]=1;fib[1]=2;for(int i=2;i<32;i++)fib[i]=fib[i-1]+fib[i-2];
    int res=0,prev=0;for(int i=30;i>=0;i--){if((n&(1<<i))!=0){res+=fib[i];if(prev==1){res--;break;}prev=1;}else prev=0;if(i==0)res++;}return res;}
    public static void main(String[] args){System.out.println(findIntegers(5));}}

File: 15-Dynamic-Programming/08-Digit-DP/cpp/digit_dp.cpp
/* Digit DP: Count numbers from 1 to N with no consecutive 1s in binary | Tags: digit-dp */
#include <bits/stdc++.h>
using namespace std;
int findIntegers(int n){vector<int>bits;while(n){bits.push_back(n&1);n>>=1;}reverse(bits.begin(),bits.end());
int sz=bits.size();vector<vector<vector<int>>>dp(sz+1,vector<vector<int>>(2,vector<int>(2,-1)));
function<int(int,int,int)>solve=[&](int pos,int prev,int tight)->int{if(pos==sz)return 1;if(dp[pos][prev][tight]!=-1)return dp[pos][prev][tight];
int limit=tight?bits[pos]:1;int res=0;for(int d=0;d<=limit;d++){if(prev==1&&d==1)continue;res+=solve(pos+1,d,tight&&(d==limit));}return dp[pos][prev][tight]=res;};
return solve(0,0,1);}
int main(){cout<<findIntegers(5)<<endl;} // 5
